erts/emulator/beam/erl_bif_binary.c

```diff
@@ -819,13 +819,31 @@ static BFReturn bm_find_first_match(BinaryFindContext *ctx, byte *haystack)
     Sint i;
     Sint j = state->pos;
     register Uint reds = *reductions;
+    byte *pos_pointer;
+    Sint needle_last = blen - 1;
+    Sint mem_read = MIN(len - j, reds);
+    ASSERT(mem_read > 0);
+
+    pos_pointer = memchr(&haystack[j], needle[needle_last], mem_read);
+    if (pos_pointer == NULL) {
+	j += mem_read;
+	reds -= mem_read;
+    } else {
+	j = pos_pointer - haystack - needle_last;
+	reds -= (pos_pointer - haystack - needle_last - j);
+    }
 
-    while (j <= len - blen) {
-	if (--reds == 0) {
+    for(;;) {
```

Comment:
I have moved this to a `for(;;)` instead of a while to make it consistent with the implementation for `bm_find_all`.